Feature/remove-BinaryFormatter

cole-brown opened the pull request:
BinaryFormatter is a security issue in .NET 5.0 - this replaces it in most places.

https://docs.microsoft.com/en-us/dotnet/standard/serialization/binaryformatter-security-guide

BinaryFormatter is replaced with BinaryReader/BinaryWriter for:
  - sysfiles
  - policies
  - ReadFile w/ PayloadMode.RawBytes 
  - WriteFile w/ PayloadMode.RawBytes

Cache files that use BinaryReader/BinaryWriter have a "magic" header number that is checked for in order to make the transition from 3.2.0 to this patch without needing to delete your file cache. The old files will just fail the new header check and the sysfile/policy/etc will continue on as if it were a failed read/missing file.

BinaryFormatter remains for:
  - ReadFile w/ PayloadMode.Serialize
  - WriteFile w/ PayloadMode.Serialize

Those would need to be serialized some other way that wasn't needed for our use case.
  - You could look into adding ZeroFormatter as a dependency: 
    - https://neuecc.medium.com/zeroformatter-fastest-c-serializer-and-infinitely-fast-deserializer-for-net-88e752803fe9
    - https://github.com/neuecc/ZeroFormatter/
    - https://www.nuget.org/packages/ZeroFormatter
  - Or possibly using an XML/JSON formatter instead:
    - https://docs.microsoft.com/en-us/dotnet/standard/serialization/binaryformatter-security-guide#preferred-alternatives

This pull request partially addresses: https://github.com/acarteas/FileCache/issues/49

cole-brown committed:
Updated FileCache to remove BinaryFormatter.

FileCacheManager:
  - Added magic header number `CACHE_VERSION`.
    + `HeaderVersionValid()` and `HeaderVersionWrite()` use it in
      writing/reading to figure out if the files were created using the
      old (BinaryFormatter) or new (BinaryWrite) way.

  - Removed `ReadSysFile()`.
    + Replaced it with:
      - `ReadSysValue<T>(string, out T)`
        + Just throws an Exception - below two functions are the
          actually used ones.
      - `ReadSysValue(string, out long)`
      - `ReadSysValue(string, out DateTime)`

  - Removed `WriteSysFile()`.
    + Replaced it with:
      - `WriteSysValue<T>(string, T)`
        + Just throws an Exception - below two functions are the
          actually used ones.
      - `WriteSysValue(string, long)`
      - `WriteSysValue(string, DateTime)`

FileCache:
  - Implemented `FileCache.PayloadMode.RawBytes` for read & write.
    + Removed code in `PayloadReadMode` property to prevent its use.
    + Implemented the `ReadFile()` function helper:
      `LoadRawPayloadData()`.
    + Updated `WriteFile()` (PayloadMode.RawBytes path) to use
      BinaryWriter instead of BinaryFormatter.

cole-brown committed:
FileCache.SerializableCacheItemPolicy fix

Removed BinaryFormatter usage from SerializableCacheItemPolicy.

SerializableCacheItemPolicy now (de)serializes via functions:
  - Serialize()
  - Deserialize()
These use BinaryWriter/BinaryReader instead of the security risk
BinaryFormatter.

SerializableCacheItemPolicy now use a magic header number like the
cached files themselves do, in order to differentiate BinaryFormatter vs
new policies.

cole-brown committed:
Cleaned up whitespace in src/FileCache files.

acarteas commented:
Cool, I'll try to take a look soon.

acarteas commented:
It looks like the some unit tests are failing when testing locally.  Can you verify?

cole-brown commented:
I did not think about the unit tests since I only pulled a subfolder into our project... 

Let me see how it goes. May just need a .NET version bumped somewhere in the project files.

cole-brown committed:
Unit test fixes.

Bumped UnitTests to netcoreapp5.0

Nuked Microsoft.SourceLink.GitHub and GitVersionTask as they were
causing build errors I didn't want to sort.

Fixed 3 of 12 failing unit tests.
  - SerializableCacheItemPolicy was not serializing its time, only date.
  - SerializableCacheItemPolicy was messing up its expiration date when
    deserializing due to the SlidingExpiration setter.

acarteas commented:
Cool, thanks for looking into it.  I haven't kept up on .net5 changes so it's good that this is getting done.

cole-brown committed:
Fixed 3 unit tests.

Fixed:
  - FileCacheTest.CacheSizeTest
  - FileCacheTest.ShrinkCacheTest
  - HashedFileCacheTest.CacheSizeTest

Moved the calculation of the new policy file's size to be outside the
'using' blocks for writing it to disk so that it could read the actual
file size instead of always getting 0.
  - It is moved entirely out of the using blocks to mirror the cache
    item's code.

Removed a duplicated line of code in FileCache.WriteHelper().
  - There was an `if` line that was there twice.
  - This has no effect on anything other than my sanity.

cole-brown committed:
Fixed FileCacheTest.CleanCacheTest

All FileCacheTest unit tests now passing.
  - 4 HashedFileCacheTest unit tests still failing.

FileCacheManager was missing the DateTime version of `WriteSysValue()`.
Added it to achieve function parity with `ReadSysValue()` functions.

cole-brown commented:
All the FileCacheTests now pass; 4 HashedFileCacheTests are still failing:
  - HashedFileCacheTest.ShrinkCacheTest
  - HashedFileCacheTest.FlushTest
  - HashedFileCacheTest.FlushRegionTest
  - HashedFileCacheTest.CleanCacheTest

Could be a pause until I get to those. New project at work and we (currently) only use BasicFileCacheManager. We are looking into using HashedFileCacheManager though.

cole-brown committed:
Simple README for the Unit Tests.

- How to run.
- Summary of suites.
- WSL2 time bug.

cole-brown committed:
Upgraded package versions.

Touched the unit testing project file.
  - Updated the nuget packages to their latest versions.
  - Added a System.Runtime.Caching 'ItemGroup Condition' for
    'netcoreapp5.0'.
  - Made a comment about the magic Service GUID.

cole-brown committed:
Fixed HashedFileCacheTest.FlushTest

Extracted FileCacheManager's code for deserializing a policy into a new
function.

Made FileCacheManager and HashedFileCacheManager use that function.

cole-brown commented:
Unit test passing again.

acarteas commented:
okay, thanks for all the work.  I'll take a look and verify over the
weekend.

On Thu, Sep 2, 2021 at 3:34 PM cole-brown ***@***.***> wrote:

> Unit test passing again.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/acarteas/FileCache/pull/52#issuecomment-912106312>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AD6Y7I2WWMFLSL6IZPWLTDTT773ZDANCNFSM5BE5NVHQ>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>

acarteas commented:
looks good, thanks a bunch for your work!

cole-brown commented:
Saw you closed 49 which this doesn't _totally_ 100% deal with. 

Copy/paste from up top: 

> BinaryFormatter remains for:
>    - ReadFile w/ PayloadMode.Serialize
>    - WriteFile w/ PayloadMode.Serialize

There wasn't an obvious solution to those and we only use raw bytes mode.

Thanks for merging my pull in and thanks for creating and maintaining FileCache!

acarteas commented:
Ah, so I guess in situations in which PayloadMode.Serialize is used, a security exception will still be thrown?  I'll reopen the ticket until I figure out a way around it.

cole-brown commented:
Yeah. It will.

In the PayloadMode.Serialize case it's not as clear what to do since you want to serialize an object/struct - the docs [suggest](https://docs.microsoft.com/en-us/dotnet/standard/serialization/binaryformatter-security-guide#preferred-alternatives) moving to XML/JSON but I'm not sure what sort of size/speed goals FileCache has.

ZeroFormatter or the like is a possibility but it requires decorating your class/struct, IIRC. 

I got this far into researching what to do about  it before I figured out we only used FileCache for storing byte arrays so using PayloadMode.RawBytes made better sense.
